File: bcweb-server/routes/amz-delete-file.js
/*
=======================================================================================================================================
API Route: amz_delete_file
=======================================================================================================================================
Method: POST
Purpose: Build the Amazon Seller Central .xlsm that DE-LISTS a set of SKUs — the "Delete file" button on the Update Amazon screen's
         reconciliation panel. It is the counterpart to POST /product-amazon, which builds the file that CREATES listings.

         The job it answers: the import's "on Amazon, unknown to us" bucket (utils/amzImport.js -> planStock, summary.stock.unknownSku)
         is SKUs the FBA inventory report shows Amazon still listing, for which we hold no skumap record. The zero-stock ones are dead
         listings cluttering the catalogue, and the only way to clear them is a Listings feed with the record action set to delete. That
         used to mean hand-editing a legacy Inventory Loader flat file per SKU.

         WHO CHOOSES THE SKUs: the client, not us. The unknown-SKU list only exists inside a just-parsed import summary — nothing is
         stashed server-side between the check and the download (see the header of the update-amazon page: an import is an action, not a
         view). So the caller posts the SKUs it is looking at. That is safe here in a way it would not be for a write route, because this
         route WRITES NOTHING: no DB connection, no amzfeed, no skumap. It is a pure formatter, and the delete only takes effect when the
         operator uploads the file to Seller Central themselves.

         ZERO-STOCK IS THE CALLER'S BAR: the client filters to afn-total-quantity = 0 before posting. Amazon will not remove a listing
         while any units — fulfillable, reserved, unsellable or inbound — sit in a fulfilment centre; the delete either bounces or leaves
         the inventory stranded. This route does not re-check that, because it has no stock data of its own: amzfeed only covers SKUs we
         know, and these are by definition the ones we do not. The bar lives with the data that can enforce it.

         Why a Python child process (identical reasoning to /product-amazon): the file is a macro-enabled .xlsm and Amazon accepts it only
         because of the settings payload embedded in the template. No Node Excel library round-trips that intact; openpyxl (keep_vba=True)
         does. The helper reuses the same in-repo SHOES.xlsm template — a delete row needs three of its 348 columns.

         SKUs go to the helper on STDIN, not argv: the list can run to hundreds and Windows caps a command line at ~32k characters.

         Delivery: standard API envelope (HTTP 200 + return_code), with the .xlsm bytes base64-encoded in `file` so the client can trigger
         a browser download — the same shape /product-amazon uses. Requires auth.
=======================================================================================================================================
Request Payload:
{
  "skus": ["AD-0XF8D-48L", "BK-7712-40"]   // string[], required, 1..1000 entries
}

Success Response:
{
  "return_code": "SUCCESS",
  "filename": "AMZ-Delete-2026-07-31.xlsm",
  "skus": 12,                // rows written (after de-duplication)
  "file": "<base64 .xlsm>"
}
=======================================================================================================================================
Return Codes:
"SUCCESS"
"MISSING_FIELDS"    // skus absent, not an array, or empty once trimmed
"NO_ROWS"           // nothing left to write after de-duplication
"TOO_MANY"          // more SKUs than the helper's safety limit for one file
"GENERATE_FAILED"   // the Python helper failed unexpectedly (missing deps/template, crash)
"UNAUTHORIZED"
"SERVER_ERROR"
=======================================================================================================================================
*/

const express = require('express');
const router = express.Router();
const path = require('path');
const os = require('os');
const fs = require('fs/promises');
const crypto = require('crypto');
const { spawn } = require('child_process');
const { verifyToken } = require('../middleware/verifyToken');
const logger = require('../utils/logger');

router.use(verifyToken);

// The Python interpreter and helper script. PYTHON_BIN overrides the default (e.g. a venv path on the VPS) — same var /product-amazon
// uses, so there is only ever one interpreter to configure.
const PYTHON_BIN = process.env.PYTHON_BIN || 'python';
const HELPER = path.join(__dirname, '..', 'scripts', 'amz-product', 'amz_delete_file.py');

// Error codes the helper emits that map straight through to a return_code (anything else -> GENERATE_FAILED).
const PASS_THROUGH = new Set(['NO_ROWS', 'TOO_MANY']);

// Guard against a client posting something enormous before we even spawn Python. The helper enforces its own limit too — this one just
// keeps a nonsense payload from reaching a child process.
const MAX_SKUS = 1000;

/** Run the helper with the SKU list on stdin, resolving with its parsed JSON stdout. Rejects with { code, message }. */
function runHelper(skus, outPath) {
  return new Promise((resolve, reject) => {
    const proc = spawn(PYTHON_BIN, [HELPER, outPath], { windowsHide: true });
    let stdout = '';
    let stderr = '';
    proc.stdout.on('data', (d) => { stdout += d; });
    proc.stderr.on('data', (d) => { stderr += d; });
    proc.on('error', (err) => reject({ code: 'GENERATE_FAILED', message: `Could not run Python helper: ${err.message}` }));

    // A write to a child that already died raises EPIPE; the 'error'/'close' handlers own the outcome, so swallow it here.
    proc.stdin.on('error', () => {});
    proc.stdin.end(JSON.stringify(skus));

    proc.on('close', (exitCode) => {
      // The helper always prints a JSON line (summary on success, {error,message} on failure). Parse the last non-empty line.
      let parsed = null;
      const line = stdout.trim().split('\n').filter(Boolean).pop();
      if (line) { try { parsed = JSON.parse(line); } catch { /* fall through to the generic error below */ } }

      if (exitCode === 0 && parsed && !parsed.error) return resolve(parsed);
      if (parsed && parsed.error) return reject({ code: parsed.error, message: parsed.message || 'Generation failed' });
      // No parseable JSON — surface stderr for the logs, generic error to the client.
      return reject({ code: 'GENERATE_FAILED', message: (stderr || stdout || 'Unknown generator error').trim() });
    });
  });
}

router.post('/', async (req, res) => {
  // Unique temp path per request so concurrent clicks never collide; cleaned up in finally.
  const outPath = path.join(os.tmpdir(), `amz-delete-${crypto.randomBytes(6).toString('hex')}.xlsm`);
  try {
    const raw = (req.body || {}).skus;
    if (!Array.isArray(raw)) {
      return res.json({ return_code: 'MISSING_FIELDS', message: 'skus must be an array' });
    }
    const skus = raw.map((s) => String(s ?? '').trim()).filter(Boolean);
    if (skus.length === 0) {
      return res.json({ return_code: 'MISSING_FIELDS', message: 'skus is empty' });
    }
    if (skus.length > MAX_SKUS) {
      return res.json({ return_code: 'TOO_MANY', message: `${skus.length} SKUs exceeds the ${MAX_SKUS} limit for one delete file` });
    }

    let summary;
    try {
      summary = await runHelper(skus, outPath);
    } catch (e) {
      if (PASS_THROUGH.has(e.code)) {
        return res.json({ return_code: e.code, message: e.message });
      }
      logger.error('[amz-delete-file] generator failed:', e.message);
      return res.json({ return_code: 'GENERATE_FAILED', message: 'Failed to build the Amazon delete file' });
    }

    const buf = await fs.readFile(outPath);
    // Dated filename: an operator may build one of these on several days running, and Downloads should not turn them into
    // "AMZ-Delete (3).xlsm". Built from the server's local date, which is the operator's day.
    const d = new Date();
    const stamp = `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}-${String(d.getDate()).padStart(2, '0')}`;
    return res.json({
      return_code: 'SUCCESS',
      filename: `AMZ-Delete-${stamp}.xlsm`,
      skus: summary.skus,
      file: buf.toString('base64'),
    });
  } catch (err) {
    logger.error('[amz-delete-file] error:', err.message);
    return res.json({ return_code: 'SERVER_ERROR', message: 'Failed to produce the Amazon delete file' });
  } finally {
    // Best-effort cleanup of the temp file (ignore if it was never created).
    fs.unlink(outPath).catch(() => {});
  }
});

module.exports = router;

File: bcweb-server/scripts/amz-product/amz_delete_file.py
"""
Amazon Listing DELETE file generator (server helper).

Builds the Seller Central .xlsm that REMOVES listings — the mirror image of amz_upload_single.py.
It exists for one job: the "on Amazon, unknown to us" bucket on the Update Amazon screen. Those are
SKUs Amazon still lists but we hold no product record for; the zero-stock ones are dead listings and
the only way to clear them is a Listings feed with the Listing Action set to delete.

Why Python and not Node (same reason as amz_upload_single.py): the file is a macro-enabled .xlsm and
Amazon accepts it only because of the settings payload in row 1 of the Template sheet. No Node Excel
library round-trips that intact; openpyxl (keep_vba=True) does.

TEMPLATE REUSE — deliberate. This writes into the SAME SHOES.xlsm that amz_upload_single.py uses,
which lives in this repo (bcweb-server/scripts/amz-product/SHOES.xlsm) and is git-tracked. A delete
row needs only three of its 348 columns, and they sit in identical positions in every Seller Central
Listings template for SHOES/UK, so a second template would be 1.1 MB of duplicate that ages on its
own. Refreshing SHOES.xlsm now fixes both flows at once.

NO DATABASE. Unlike amz_upload_single.py this touches nothing — it does not connect to Postgres and
writes no rows. It is a pure formatter: SKUs in, file out. That is the point. The SKUs it is given
are by definition unknown to our database, so there is nothing here to stamp; and a delete is applied
by Amazon when the operator uploads the file, not by us. Keeping it read-nothing/write-nothing means
clicking the button can never leave our data in a state that assumes an upload that never happened.

Input is a JSON array of SKU strings on STDIN, not argv: the list can run to hundreds of entries and
Windows caps a command line at ~32k characters.

Usage:
    echo '["SKU-1","SKU-2"]' | python amz_delete_file.py <OUTPUT_XLSM_PATH>

Exit codes: 0 = success (JSON summary on stdout); non-zero = failure (JSON error on stdout).
"""

import os
import sys
import json
import shutil

import openpyxl

# The template lives next to this script — same copy amz_upload_single.py uses.
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
TEMPLATE_PATH = os.path.join(SCRIPT_DIR, "SHOES.xlsm")

# Column positions in the SHOES.xlsm Template sheet (openpyxl is 1-based). A delete needs only these
# three; every other column must stay EMPTY or Amazon treats the row as a listing update instead.
#   A = contribution_sku#1.value   B = product_type#1.value   C = ::record_action
COL_SKU = 1
COL_PRODUCT_TYPE = 2
COL_LISTING_ACTION = 3

# Data rows start at row 7 (1 = settings blob, 3-4 = headings/labels, 5 = attribute names, 6 = examples).
DATA_START_ROW = 7

# The product type the template was generated for. Not a free choice — it must match the template's
# own settings payload, and SHOES.xlsm is the SHOES/UK template.
PRODUCT_TYPE = "SHOES"

# ::record_action value for a removal. The raw enum, not the "Delete" dropdown label, to match
# amz_upload_single.py which writes "partial_update" the same way and is accepted in production.
RECORD_ACTION_DELETE = "delete"

# Safety rail. The unknown-SKU bucket is normally a handful to a few hundred; a four-figure delete file
# means something upstream is wrong (a truncated or mis-parsed inventory report), and refusing is a lot
# cheaper than an operator uploading a file that de-lists the catalogue.
MAX_SKUS = 1000


def fail(code, message):
    """Emit a JSON error envelope on stdout and exit non-zero for the Node route to map to a return_code."""
    print(json.dumps({"error": code, "message": message}))
    sys.exit(1)


def generate_delete_file(skus, output_path):
    """Copy the template and write one delete row per SKU, saving to output_path."""
    shutil.copy2(TEMPLATE_PATH, output_path)
    wb = openpyxl.load_workbook(output_path, keep_vba=True)
    ws = wb["Template"]

    # Clear any example/left-over data rows the template shipped with, so the file contains our rows
    # and nothing else. Mirrors amz_upload_single.py; only the three columns we write can be dirty.
    for r in range(DATA_START_ROW, DATA_START_ROW + 1000):
        if ws.cell(row=r, column=COL_SKU).value is None:
            break
        for col in (COL_SKU, COL_PRODUCT_TYPE, COL_LISTING_ACTION):
            ws.cell(row=r, column=col, value=None)

    for i, sku in enumerate(skus):
        r = DATA_START_ROW + i
        ws.cell(row=r, column=COL_SKU, value=sku)
        ws.cell(row=r, column=COL_PRODUCT_TYPE, value=PRODUCT_TYPE)
        ws.cell(row=r, column=COL_LISTING_ACTION, value=RECORD_ACTION_DELETE)

    wb.save(output_path)


def main():
    if len(sys.argv) != 2:
        fail("BAD_ARGS", "Usage: amz_delete_file.py <OUTPUT_XLSM_PATH>  (SKU JSON array on stdin)")

    output_path = sys.argv[1]
    if not os.path.exists(TEMPLATE_PATH):
        fail("NO_TEMPLATE", f"Template not found at {TEMPLATE_PATH}")

    try:
        payload = json.load(sys.stdin)
    except Exception as e:
        fail("BAD_ARGS", f"Could not read the SKU list from stdin: {e}")

    if not isinstance(payload, list):
        fail("BAD_ARGS", "Expected a JSON array of SKU strings on stdin")

    # De-duplicate while preserving order: two rows for one SKU is a rejected feed, and the caller
    # builds this list from a report we do not control.
    seen = set()
    skus = []
    for raw in payload:
        sku = str(raw).strip()
        if not sku or sku in seen:
            continue
        seen.add(sku)
        skus.append(sku)

    if not skus:
        fail("NO_ROWS", "No SKUs to delete")
    if len(skus) > MAX_SKUS:
        fail("TOO_MANY", f"{len(skus)} SKUs exceeds the {MAX_SKUS} safety limit for one delete file")

    generate_delete_file(skus, output_path)
    print(json.dumps({"skus": len(skus)}))


if __name__ == "__main__":
    main()
